### PR TITLE
feat: Automated Testing for OFM Takeover Tool

### PR DESCRIPTION
**Pull request by cmaddox5**
**Asana task**: [OFM Takeover periodic tests/monitoring](https://app.asana.com/0/1164574158255689/1206301091214212/f)

Due to the Takeover Tool issues we've had in the last 12 months, this automated test will ensure that directory structure is what we expect and that we can upload/delete images to the test folder OFM provided. This test folder is the only folder that images can be uploaded to in this job. OFM confirmed that this folder is not configured for any screen so we don't have to worry about impacting live screens.

### CODE REVIEW

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (digitalcora) [inline lib/screenplay/jobs/takeover_tool_testing_job.ex:36]: We use these specific directory names both here and I think twice in `all_directories_exist` — could we extract them to constants? It's a small thing, but I'd want to make sure that if Outfront decides to rename these, we can change the name in one place and be confident we got everything.

```elixir
@landscape_dir "Landscape"
@portrait_dir "Portrait"
```

Following that train of thought... are these also hard-coded in the "real" tool?

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (digitalcora) [inline lib/screenplay/jobs/takeover_tool_testing_job.ex:33]: I think we could compile this image data into the app as a module attribute, and then the app wouldn't compile if the image wasn't present as expected (instead of crashing later when the job runs). Freehanded so idk if this exact code works:

```elixir
@test_image :code.priv_dir(:screenplay) |> Path.join("takeover_test.png") |> File.read!
```

Technically we should also specify this as an [`@external_resource`](https://hexdocs.pm/elixir/Module.html#module-external_resource) of the module, though it's unlikely to change much.

File: assets/js/components/OutfrontTakeoverTool/OutfrontTakeoverTool.tsx
Comment (digitalcora) [inline assets/js/components/OutfrontTakeoverTool/OutfrontTakeoverTool.tsx:75]: What's the reason we have to specify this both here and on L171?

File: lib/screenplay/outfront/sftp.ex
Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex]: Yep, totally. Was able to move those to the app config and removed the old module attribute.

File: .gitignore
Comment (cmaddox5) [reply .gitignore:43]: Great idea. I'll move all that around and change the code and README setup instructions.

File: lib/screenplay_web/controllers/outfront_takeover_tool/page_controller.ex
Comment (cmaddox5) [reply lib/screenplay_web/controllers/outfront_takeover_tool/page_controller.ex:27]: Good call. Default was definitely overly cautious.

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (cmaddox5) [reply lib/screenplay/jobs/takeover_tool_testing_job.ex:121]: Oops. I did that with this [anti-pattern](https://hexdocs.pm/elixir/1.16.1/code-anti-patterns.html#complex-extractions-in-clauses) in mind, but I actually misunderstood the main concern of the anti-pattern. Will get this changed.

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (cmaddox5) [reply lib/screenplay/jobs/takeover_tool_testing_job.ex:36]: They are hard-coded, yes. OFM's SFTP directory structure is `./Landscape/...` and `./Portrait/...`.

File: assets/js/components/OutfrontTakeoverTool/OutfrontTakeoverTool.tsx
Comment (cmaddox5) [reply assets/js/components/OutfrontTakeoverTool/OutfrontTakeoverTool.tsx:75]: There isn't one. I think this is leftover from when I was passing this state down via `props` instead of using the context. It will still need a default of `{}`, but no need to provide the full object shape in it.

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (cmaddox5) [reply lib/screenplay/jobs/takeover_tool_testing_job.ex:36]: I think I misunderstood what you meant by "real" tool for some reason. They were indeed hardcoded over there are well. I made those module constants as well, but am not a fan of having two sets of constants that refer to the same thing. What would be the best way for these two modules to share these constants? App config maybe?

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (cmaddox5) [reply lib/screenplay/jobs/takeover_tool_testing_job.ex:33]: Hmm, that's a cool idea. Love it.

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (cmaddox5) [reply lib/screenplay/jobs/takeover_tool_testing_job.ex:135]: Yep, definitely could be compile-time.

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (cmaddox5) [reply lib/screenplay/jobs/takeover_tool_testing_job.ex]: I added a new function `SFTP.run/1` to help with this. Is this what you had in mind?

File: mix.exs
Comment (cmaddox5) [reply mix.exs:61]: It currently has a single instance. And honestly, I think it will only have a single instance for a long time. The userbase for Screenplay is very small so traffic should always be light enough so ECS does not scale us to use another task. In the next feature for Screenplay, we will likely be using a psql DB. When that happens, we could consider switching to [Oban](https://github.com/sorentwo/oban) for job processing. It has a feature that prevents duplicate jobs from running.

File: lib/screenplay_web/router.ex
Comment (digitalcora) [inline lib/screenplay_web/router.ex:97]: It looks like this is already in a `ScreenplayWeb.OutfrontTakeoverTool` scope, so is it redundant to specify `outfront_takeover_tool` in the path / action name? (though confusingly it shares an `/api` path prefix with other, non-OFTT routes...)

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (digitalcora) [inline lib/screenplay/jobs/takeover_tool_testing_job.ex:27]: I think we no longer need this skip, since we're not doing the file reading at runtime.

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (digitalcora) [inline lib/screenplay/jobs/takeover_tool_testing_job.ex:18]: If this is not allowed to be nil, we should use the shoutier `compile_env!`

File: lib/screenplay/outfront/sftp.ex
Comment (digitalcora) [inline lib/screenplay/outfront/sftp.ex:12]: ...and here

File: lib/screenplay/outfront/sftp.ex
Comment (digitalcora) [inline lib/screenplay/outfront/sftp.ex:19]: I'm wondering about the purpose of this `rescue`. Is there a reason we shouldn't either "let it crash" or return an error to the caller? Generally, it's a good idea to avoid silently swallowing errors (from the caller's perspective, even if the error is logged somewhere as a side effect).

File: lib/screenplay/outfront/sftp.ex
Comment (digitalcora) [inline lib/screenplay/outfront/sftp.ex:45]: Nice, it looks like we're gaining some better cleanup of the connection by wrapping these ✨

File: lib/screenplay/jobs/takeover_tool_testing_job.ex
Comment (digitalcora) [reply lib/screenplay/jobs/takeover_tool_testing_job.ex]: Maybe that new config file you extracted, since we are already storing some other directory names in there? They could then still be module constants, just loaded with `compile_env!`.

File: lib/screenplay/outfront/sftp.ex
Comment (digitalcora) [inline lib/screenplay/outfront/sftp.ex:5]: What were the errors here? I don't have this repo set up on my machine yet >.>

**Review by digitalcora — COMMENTED** [review]
Looking good, just a few minor things I noticed. Can we test this out in a real deployed environment? (by changing the Quantum conditional to `dev-green` maybe)

File: lib/screenplay/outfront/sftp.ex
Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex:5]: `Pattern will never match` errors for the error cases in these functions. The `FakeSFTPClient` doesn't really any errors so when dialyzer complains while that is the module configured for local development.

Comment (digitalcora) [reply lib/screenplay/outfront/sftp.ex:5]: Hmm... I wonder if there's a way to fix that while preserving the warnings for the case where we actually do forget to handle a match. If we explicitly spec the function in the SFTP client as returning `:ok | {:error, ...}`, does that then result in an overspec error?

File: lib/screenplay/outfront/sftp.ex
Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex:19]: The main purpose was to log to Sentry. The big gotcha that I missed is we don't have Sentry logging for server errors at the moment 🙃 . That's not great cause that was what I assumed we would use for getting notified of these issues... I'll remove the `rescue` regardless because if we set up Sentry correctly, error logs will show up there automatically.

Comment (digitalcora) [reply lib/screenplay/outfront/sftp.ex:19]: Good thing we tested it 😅 Will you set that up as part of this task? Seems pretty essential to the "feature".

Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex:19]: Yeah, I think it's important enough to be in scope for this task. Will get started on that and make sure it's working before merging this branch.

File: lib/screenplay/outfront/sftp.ex
Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex:5]: Unfortunately, dialyzer seems to ignore the spec and looks at what the function actually returns. So I would either need to fake an error return for these or ignore the warnings.

Comment (digitalcora) [reply lib/screenplay/outfront/sftp.ex:5]: > Unfortunately, dialyzer seems to ignore the spec and looks at what the function actually returns

Well, normally I'd say that's _fortunate_ behavior, but 😅

I wonder why this hasn't been a problem in the past when I've had fake versions of services that e.g. always returned `:ok`. I think the setup I've used was where the real and fake versions shared a `@behaviour`? But anyway, I don't think it's super important to get to the bottom of this right now unless we've got extra time.

File: lib/screenplay/outfront/sftp.ex
Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex]: Looks like I was fooled by how Screenplay was set up in Sentry. The platform selected in the Project Settings was React so I assumed it was only set up for the frontend. We do have the necessary back configuration for server errors as well. Only thing missing was a separate DSN for dev-green (was still using the dev DSN. PR for that change [here](https://github.com/mbta/devops/pull/1639)). I changed the platform and project name in Sentry so it is set up the way Screens is (which we seem happy with so a good standard for Screenplay).

Comment (digitalcora) [reply lib/screenplay/outfront/sftp.ex]: Wait, it needs a separate DSN? I thought the environment name was just configured in the app and sent along with error reports. I don't remember having to set up more than one DSN per project.

Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex]: Very interesting. For whatever reason, I thought that each environment needed their own DSN when I initially set up Sentry for our apps. You're right though, we can use the `environment_name` to identify environment and set up alerts based on that. I closed that devops PR. Not going to mess with existing DSNs for now.

File: lib/screenplay/outfront/sftp.ex
Comment (cmaddox5) [reply lib/screenplay/outfront/sftp.ex:5]: Funnily enough, I did try using a `behaviour` and it still yelled about it. I wonder if it's because the "real" module we are using is not ours. I'm not entirely sure, but I'm not super worried about ignoring the warnings specifically for these functions.

**Review by digitalcora — APPROVED** [review]
Presuming this has been tested in a deployed environment, lgtm!

**Pull request merged**
